core/src/main/java/org/apache/calcite/plan/RelOptUtil.java

```diff
@@ -2199,13 +2199,49 @@ public static boolean equal(
       RelDataType type2,
       Litmus litmus) {
     if (!areRowTypesEqual(type1, type2, false)) {
-      return litmus.fail("Type mismatch:\n{}:\n{}\n{}:\n{}",
+      return litmus.fail("Type mismatch:\n{}:\n{}\n{}:\n{}\ndifference:\n{}",
           desc1, type1.getFullTypeString(),
-          desc2, type2.getFullTypeString());
+          desc2, type2.getFullTypeString(),
+          getFullTypeDifferenceString(type1, type2));
     }
     return litmus.succeed();
   }
 
+  public static String getFullTypeDifferenceString(RelDataType source,
+      RelDataType target) {
+    if (source == target) {
+      return "";
+    }
+
+    if (source.getFieldCount() > target.getFieldCount()) {
+      return "Type mismatch: the target type has less fields than the source type";
+    } else if (source.getFieldCount() < target.getFieldCount()) {
+      return "Type mismatch: the target type has more fields than the source type";
+    }
+
+    final StringBuilder stringBuilder = new StringBuilder("Type mismatch: \n");
```

Comment:
If the fields are identical, we should not have the prefix "Type mismatch"?

Reply:
Makes sense.